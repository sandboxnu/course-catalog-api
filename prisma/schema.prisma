generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL_WITH_CONNECTIONS")
}

model Course {
  classAttributes String[]         @map("class_attributes")
  classId         String?          @map("class_id")
  coreqs          Json?
  description     String?
  feeAmount       Int?             @map("fee_amount")
  feeDescription  String?          @map("fee_description")
  host            String?
  id              String           @id
  lastUpdateTime  DateTime?        @default(now()) @map("last_update_time")
  maxCredits      Int?             @map("max_credits")
  minCredits      Int?             @map("min_credits")
  name            String?
  nupath          String[]
  optPrereqsFor   Json?            @map("opt_prereqs_for")
  prereqs         Json?
  prereqsFor      Json?            @map("prereqs_for")
  prettyUrl       String?          @map("pretty_url")
  subject         String?
  termId          String?          @map("term_id")
  url             String?
  FollowedCourse  FollowedCourse[]
  sections        Section[]

  @@unique([classId, termId, subject], name: "uniqueCourseProps", map: "uniqueCourseProps")
  @@map("courses")
}

model Major {
  majorId      String @map("major_id")
  plansOfStudy Json   @map("plans_of_study")
  yearVersion  String @map("year_version")
  spec         Json

  @@id([yearVersion, majorId])
  @@map("majors")
}

model Professor {
  id                String  @id
  name              String
  firstName         String  @map("first_name")
  lastName          String  @map("last_name")
  primaryDepartment String? @map("primary_department")
  primaryRole       String? @map("primary_role")
  phone             String?
  email             String?
  officeRoom        String? @map("office_room")

  @@map("professors")
}

model Section {
  classHash       String?           @map("class_hash")
  classType       String?           @map("class_type")
  crn             String?
  honors          Boolean?
  id              String            @id
  info            String?
  meetings        Json?
  campus          String?
  profs           String[]
  seatsCapacity   Int?              @map("seats_capacity")
  seatsRemaining  Int?              @map("seats_remaining")
  url             String?
  waitCapacity    Int?              @map("wait_capacity")
  waitRemaining   Int?              @map("wait_remaining")
  lastUpdateTime  DateTime?         @default(now()) @map("last_update_time")
  course          Course?           @relation(fields: [classHash], references: [id], onDelete: Cascade)
  FollowedSection FollowedSection[]

  @@map("sections")
}

model Subject {
  abbreviation String @id
  description  String

  @@map("subjects")
}

model FollowedCourse {
  courseHash String @map("course_hash")
  userId     Int    @map("user_id")
  course     Course @relation(fields: [courseHash], references: [id], onDelete: Cascade)
  user       User   @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@id([userId, courseHash])
  @@map("followed_courses")
}

model FollowedSection {
  sectionHash String  @map("section_hash")
  userId      Int     @map("user_id")
  section     Section @relation(fields: [sectionHash], references: [id], onDelete: Cascade)
  user        User    @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@id([userId, sectionHash])
  @@map("followed_sections")
}

model User {
  id               Int               @id @default(autoincrement())
  phoneNumber      String            @unique(map: "users.phone_number_unique") @map("phone_number")
  FollowedCourse   FollowedCourse[]
  followedSections FollowedSection[]

  @@map("users")
}

model TermInfo {
  termId     String @unique(map: "term_ids.term_id_unique") @map("term_id")
  subCollege String @map("sub_college")
  text       String
  active     Boolean @default(true)

  @@map("term_ids")
}
