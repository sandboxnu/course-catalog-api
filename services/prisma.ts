import { PrismaClient } from "@prisma/client";
import macros from "../utils/macros";

let prisma: PrismaClient;
try {
  prisma = new PrismaClient({
    log: ["info", "warn", "error"],
  });
  macros.log("** Created Prisma client");

  // TEMP / TODO - REMOVE / DO NOT LEAVE HERE PLEASE
  // Temp fix to address Prisma connection pool issues
  // https://github.com/prisma/prisma/issues/7249#issuecomment-1059719644
  const intervalTime = 6 * 60 * 60_000; // Every 6 hours
  setInterval(async () => {
    const startTime = Date.now();
    await prisma.$disconnect();
    macros.log("Disconnected Prisma");
    await prisma.$connect();
    const totalTime = Date.now() - startTime;
    macros.log(
      `Reconnected Prisma - downtime of ${totalTime} ms (${
        totalTime / 60_000
      } mins)`
    );
  }, intervalTime);
} catch (e) {
  macros.error(e);
}

export default prisma;
