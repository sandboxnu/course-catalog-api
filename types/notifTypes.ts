export interface UserInfo {
  phoneNumber: string;
  courseIds: string[];
  sectionIds: string[];
}

// Stores information for all changes to a course or section
export interface NotificationInfo {
  updatedCourses: CourseNotificationInfo[];
  updatedSections: SectionNotificationInfo[];
}

// marks new sections being added to a Course
export interface CourseNotificationInfo {
  subject: string;
  courseId: string;
  termId: string;
  courseHash: string;
  numberOfSectionsAdded: number;
  campus: string;
}

// marks seats becoming available in a section
export interface SectionNotificationInfo {
  subject: string;
  courseId: string;
  sectionHash: string;
  termId: string;
  seatsRemaining: number;
  crn: string;
  campus: string;
}
