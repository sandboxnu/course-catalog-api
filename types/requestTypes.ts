import { CoreOptions } from "request";
import http from "http";
import https from "https";

export interface HostAnalytics {
  totalBytesDownloaded: number;
  totalErrors: number;
  totalGoodRequests: number;
  startTime: number | null;
}

export type RequestAnalytics = Record<string, HostAnalytics>;

export interface RequestPool {
  maxSockets: number;
  keepAlive: boolean;
  maxFreeSockets: number;
  "https:false:ALL"?: https.Agent;
  "http:"?: http.Agent;
}

export interface AgentAnalytics {
  socketCount: number;
  requestCount: number;
  maxSockets: number;
}

export interface CustomRequestConfig extends CoreOptions {
  url: string;
  method: string;
  cache?: boolean;
  cacheName?: string;
  headers?: Record<string, string>;
  pool?: RequestPool;
}

export type PartialRequestConfig =
  | (Partial<CustomRequestConfig> & { url: string })
  | string;

export interface NativeRequestConfig extends CustomRequestConfig {
  method: string;
  headers: Record<string, string>;
  pool: RequestPool;
  timeout: number;
  resolveWithFullResponse: boolean;
  rejectUnauthorized: boolean;
  requestCert: boolean;
  ciphers: string;
}

export type AmplitudeEvent = Partial<AgentAnalytics> & {
  totalBytesDownloaded?: number;
  totalErrors?: number;
  totalGoodRequests?: number;
  startTime?: number | null;
  hostname: string;
};

export type DoRequestReturn = { items: unknown; totalCount: number };
