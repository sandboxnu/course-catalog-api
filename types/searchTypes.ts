/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import { Employee, Course, Section } from "./types";

/****************** UTILS ******************/
type OneOrMany<T> = T | T[];

/***************** QUERIES *****************/
export interface EsQuery {
  from: number;
  size: number;
  sort: EsSort;
  query: QueryNode;
  aggregations?: QueryAgg;
}

export type EsValue = string | string[] | boolean | Range;

export interface Range {
  min: number;
  max: number;
}

// every query is either a leaf query or a bool query
export type QueryNode = LeafQuery | BoolQuery;

// ====== Leaf Queries =======
export type LeafQuery =
  | TermQuery
  | TermsQuery
  | ExistsQuery
  | MultiMatchQuery
  | RangeQuery
  | MatchAllQuery
  | BoolQuery;

export interface TermQuery {
  term: FieldQuery;
}

export interface TermsQuery {
  terms: FieldQuery;
}

export interface MultiMatchQuery {
  multi_match: {
    query: string;
    type: string;
    fields: string[];
  };
}

export interface RangeQuery {
  range: {
    [fieldName: string]: {
      gte?: number;
      gt?: number;
      lte?: number;
      lt?: number;
    };
  };
}

export interface ParsedQuery {
  phraseQ: LeafQuery[];
  fieldQ: LeafQuery;
}

export const MATCH_ALL_QUERY = { match_all: {} };
export type MatchAllQuery = typeof MATCH_ALL_QUERY;

export interface ExistsQuery {
  exists: { field: string };
}

export interface FieldQuery {
  [fieldName: string]: EsValue;
}
// ====== Bool Queries =======
export interface BoolQuery {
  bool: BoolType;
}

export type BoolType = MustQuery | ShouldQuery | FilterQuery;

/**
 * The below three Query types have the same structure.
 * The only difference is what string shows up first.
 * But, they need to be separate types, because they mean very different things.
 * Not sure how to solve.
 */
export interface MustQuery {
  must: OneOrMany<QueryNode>;
}

export interface ShouldQuery {
  should: OneOrMany<QueryNode>;
  minimum_should_match?: number;
}

export interface FilterQuery {
  filter: OneOrMany<QueryNode>;
}

// ====== Misc. Queries ======
export type EsSort = [SortType, SortInfo];
type SortType = string;

export interface SortInfo {
  [fieldName: string]: {
    order: string;
    unmapped_type: string;
  };
}

export interface QueryAgg {
  [aggName: string]: {
    terms: {
      field: string;
      size?: number;
    };
  };
}

// ========= Filters =========
export type FilterInput = {
  [filterName: string]: EsValue;
};

interface FilterStruct<Input, T extends Input = Input> {
  validate(input: unknown): input is T;
  create(input: T): LeafQuery;
  agg: AggProp;
}

export type EsFilterStruct<T extends EsValue = EsValue> = FilterStruct<
  EsValue,
  T
>;

export type FilterPrelude = Record<string, EsFilterStruct>;

// ======= Agg Filters =======
export type AggProp = false | string;

interface AggFilterStruct<Input> extends FilterStruct<Input> {
  agg: string;
}

export type EsAggFilterStruct = AggFilterStruct<EsValue>;

export type AggFilterPrelude = Record<string, EsAggFilterStruct>;

/*************** ES RESULTS ****************/
// TODO blocked a bit by new ORM, trying to avoid duplication
export type EsResult = any;

export type EsMultiResult = {
  body: {
    responses: EsResultBody[];
  };
};

export interface EsResultBody {
  took: number;
  hits: {
    total: { value: number };
    hits: EsResult[];
    value: number;
  };
  aggregations: EsAggResults;
}

export interface EsAggResults {
  [aggName: string]: {
    buckets: Array<{ key: string; doc_count: number }>;
  };
}

/************* SEARCH RESULTS **************/
export interface SearchResults {
  searchContent: SearchResult[];
  resultCount: number;
  took: {
    total: number;
    hydrate: number;
    es: number;
  };
  aggregations: AggResults;
}

export interface PartialResults {
  output: EsResult[];
  took: number;
  aggregations: AggResults;
}

export interface SingleSearchResult {
  results: SearchResult[];
  resultCount: number;
  took: number;
  hydrateDuration: number;
  aggregations: AggResults;
}

export interface AggCount {
  value: string;
  count: number;
  description?: string;
}

export interface AggResults {
  [filterName: string]: AggCount[];
}

export type CourseSearchResult = {
  type: "class";
  class: Course;
  sections: Section[];
};
export type ProfessorSearchResult = { type: "employee"; employee: Employee };

export type SearchResult = CourseSearchResult | ProfessorSearchResult;

/**************** ES UTILS *****************/
export type EsMapping = Record<string, any>;

// TODO value is a Document--something we insert into ES
export type EsBulkData = Record<string, any>;
