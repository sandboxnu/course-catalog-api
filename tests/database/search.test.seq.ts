import searcher from "../../services/searcher";
import prisma from "../../services/prisma";
import { LeafQuery, ParsedQuery } from "../../types/searchTypes";

beforeAll(async () => {
  searcher.subjects = {};
});

describe("searcher", () => {
  describe("generateMQuery", () => {
    it("generates with no filters", () => {
      expect(
        searcher.generateMQuery("fundies", "202030", 0, 10, {})
      ).toMatchSnapshot();
    });

    it("generates aggs with campus filters applied", () => {
      expect(
        searcher.generateMQuery("fundies", "202030", 0, 10, {
          campus: ["Online", "Boston"],
        })
      ).toMatchSnapshot();
    });
  });

  //Unit tests for the parseQuery function
  describe("parseQuery", () => {
    it("query with no phrases", () => {
      const retQueries: ParsedQuery = searcher.parseQuery(
        "this is a query with no phrases"
      );
      expect(retQueries.phraseQ.length).toEqual(0); //no phrase queries
      expect(retQueries.fieldQ).not.toEqual(null);
      const fieldQuery: LeafQuery = retQueries.fieldQ;

      expect(fieldQuery).toEqual({
        multi_match: {
          query: "this is a query with no phrases",
          type: "most_fields",
          fields: searcher.getFields(),
        },
      });
    });

    it("query with just a phrase", () => {
      const retQueries: ParsedQuery = searcher.parseQuery('"this is a phrase"');

      expect(retQueries.phraseQ.length).toEqual(1);
      expect(retQueries.fieldQ).toEqual(null);

      const phraseQuery: LeafQuery = retQueries.phraseQ[0];

      expect(phraseQuery).toEqual({
        multi_match: {
          query: "this is a phrase",
          type: "phrase",
          fields: searcher.getFields(),
        },
      });
    });

    it("query with a phrase and other text", () => {
      const retQueries: ParsedQuery = searcher.parseQuery('text "phrase" text');
      expect(retQueries.phraseQ.length).toEqual(1);
      expect(retQueries.fieldQ).not.toEqual(null);

      const phraseQuery: LeafQuery = retQueries.phraseQ[0];
      const fieldQuery: LeafQuery = retQueries.fieldQ;

      expect(phraseQuery).toEqual({
        multi_match: {
          query: "phrase",
          type: "phrase",
          fields: searcher.getFields(),
        },
      });

      expect(fieldQuery).toEqual({
        multi_match: {
          query: "text text",
          type: "most_fields",
          fields: searcher.getFields(),
        },
      });
    });
  });
  // TODO: create an association between cols in elasticCourseSerializer and here
  describe("generateQuery", () => {
    it("generates match_all when no query", () => {
      expect(
        searcher.generateQuery("", "202030", {}, 0, 10).query["bool"]["must"]
      ).toEqual({ match_all: {} });
    });

    it("generates a query without filters", () => {
      expect(
        searcher.generateQuery("fundies", "202030", {}, 0, 10, "nupath")
      ).toMatchSnapshot();
    });
  });

  describe("validateFilters", () => {
    it("removes invalid filters", () => {
      const invalidFilters = {
        NUpath: "NU Core/NUpath Adv Writ Dscpl",
        college: "GS Col of Arts",
        subject: "CS",
        online: false,
        campus: ["Boston"],
        classType: ["Lecture"],
        inValidFilterKey: "",
      };
      expect(searcher.validateFilters(invalidFilters)).toMatchObject({});
    });

    it("keeps all valid filters", () => {
      const validFilters = {
        nupath: [
          "NU Core/NUpath Adv Writ Dscpl",
          "NUpath Interpreting Culture",
        ],
        subject: ["ENGW", "ARTG", "CS"],
        campus: ["Boston"],
        classType: ["Lecture"],
      };
      expect(searcher.validateFilters(validFilters)).toMatchObject(
        validFilters
      );
    });
  });

  describe("Single search result", () => {
    beforeEach(async () => {
      await prisma.section.deleteMany({});
      await prisma.course.deleteMany({});
      await prisma.course.create({
        data: {
          id: "neu.edu/202030/CS/2500",
          host: "neu.edu",
          classId: "2500",
          name: "Fundamentals of Computer Science 1",
          termId: "202030",
          subject: "CS",
          lastUpdateTime: new Date(),
        },
      });
      await prisma.course.create({
        data: {
          id: "neu.edu/202030/PHIL/1145",
          host: "neu.edu",
          classId: "1145",
          name: "Tech and Human Values",
          termId: "202030",
          subject: "PHIL",
          nupath: { set: ["Ethical reasoning", "Argue", "Live in the mud"] },
        },
      });
      await prisma.course.create({
        data: {
          id: "neu.edu/202030/CS/2510",
          host: "neu.edu",
          classId: "2510",
          name: "Fundamentals of Computer Science 2",
          termId: "202030",
          subject: "CS",
        },
      });
      await prisma.section.create({
        data: {
          id: "neu.edu/202030/CS/2500/19350",
          course: { connect: { id: "neu.edu/202030/CS/2500" } },
          seatsCapacity: 80,
          seatsRemaining: 0,
          classType: "Lecture",
          campus: "Seattle, WA",
          honors: false,
        },
      });
      await prisma.section.create({
        data: {
          id: "neu.edu/202030/PHIL/1145/20142",
          course: { connect: { id: "neu.edu/202030/PHIL/1145" } },
          seatsCapacity: 40,
          seatsRemaining: 0,
          classType: "Lecture",
          campus: "Boston",
          honors: true,
        },
      });
    });
    describe("getSingleResultAggs", () => {
      it("Gets aggregation for single result", async () => {
        const singleResult = await prisma.course.findUnique({
          where: { id: "neu.edu/202030/CS/2500" },
          include: { sections: true },
        });
        expect(searcher.getSingleResultAggs(singleResult)).toEqual({
          campus: [{ value: "Seattle, WA", count: 1 }],
          nupath: [],
          subject: [{ value: "CS", count: 1 }],
          classType: [{ value: "Lecture", count: 1 }],
          honors: [{ value: "false", count: 1 }],
        });
      });
      it("Gets aggregation for single result with nupath", async () => {
        const singleResult = await prisma.course.findUnique({
          where: { id: "neu.edu/202030/PHIL/1145" },
          include: { sections: true },
        });
        expect(searcher.getSingleResultAggs(singleResult)).toEqual({
          campus: [{ value: "Boston", count: 1 }],
          nupath: [
            { value: "Ethical reasoning", count: 1 },
            { value: "Argue", count: 1 },
            { value: "Live in the mud", count: 1 },
          ],
          subject: [{ value: "PHIL", count: 1 }],
          classType: [{ value: "Lecture", count: 1 }],
          honors: [{ value: "true", count: 1 }],
        });
      });
    });
    describe("getOneSearchResult", () => {
      it("Gets results for valid class", async () => {
        expect(
          await searcher.getOneSearchResult("CS", "2500", "202030")
        ).toMatchObject({
          results: [
            {
              class: {
                id: "neu.edu/202030/CS/2500",
                host: "neu.edu",
                classId: "2500",
                name: "Fundamentals of Computer Science 1",
                termId: "202030",
                subject: "CS",
                lastUpdateTime: expect.anything(),
              },
            },
          ],
          resultCount: 1,
          took: 0,
          hydrateDuration: expect.anything(),
          aggregations: {
            nupath: [],
            subject: [{ count: 1, value: "CS" }],
            classType: [{ count: 1, value: "Lecture" }],
          },
        });
      });
      it("Gets 0 results for invalid course", async () => {
        expect(
          await searcher.getOneSearchResult("CS", "2504", "202030")
        ).toMatchObject({
          results: [],
          resultCount: 0,
          took: 0,
          hydrateDuration: expect.anything(),
          aggregations: {
            nupath: [],
            subject: [],
            classType: [],
          },
        });
      });
      it("Gets 0 results for course with no sections", async () => {
        expect(
          await searcher.getOneSearchResult("CS", "2510", "202030")
        ).toMatchObject({
          results: [],
          resultCount: 0,
          took: 0,
          hydrateDuration: expect.anything(),
          aggregations: {
            nupath: [],
            subject: [],
            classType: [],
          },
        });
      });
    });
  });
});
