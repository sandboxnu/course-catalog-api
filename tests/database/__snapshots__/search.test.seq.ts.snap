// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searcher generateMQuery generates aggs with campus filters applied 1`] = `
[
  {
    "aggregations": undefined,
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                    {
                      "terms": {
                        "sections.campus.keyword": [
                          "Online",
                          "Boston",
                        ],
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 10,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  {
    "aggregations": {
      "nupath": {
        "terms": {
          "field": "class.nupath.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                    {
                      "terms": {
                        "sections.campus.keyword": [
                          "Online",
                          "Boston",
                        ],
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  {
    "aggregations": {
      "subject": {
        "terms": {
          "field": "class.subject.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                    {
                      "terms": {
                        "sections.campus.keyword": [
                          "Online",
                          "Boston",
                        ],
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  {
    "aggregations": {
      "classType": {
        "terms": {
          "field": "sections.classType.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                    {
                      "terms": {
                        "sections.campus.keyword": [
                          "Online",
                          "Boston",
                        ],
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  {
    "aggregations": {
      "campus": {
        "terms": {
          "field": "sections.campus.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              {
                "term": {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  {
    "aggregations": {
      "honors": {
        "terms": {
          "field": "sections.honors",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                    {
                      "terms": {
                        "sections.campus.keyword": [
                          "Online",
                          "Boston",
                        ],
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
]
`;

exports[`searcher generateMQuery generates with no filters 1`] = `
[
  {
    "aggregations": undefined,
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              {
                "term": {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 10,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  {
    "aggregations": {
      "nupath": {
        "terms": {
          "field": "class.nupath.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              {
                "term": {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  {
    "aggregations": {
      "subject": {
        "terms": {
          "field": "class.subject.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              {
                "term": {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  {
    "aggregations": {
      "classType": {
        "terms": {
          "field": "sections.classType.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              {
                "term": {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  {
    "aggregations": {
      "campus": {
        "terms": {
          "field": "sections.campus.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              {
                "term": {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  {
    "aggregations": {
      "honors": {
        "terms": {
          "field": "sections.honors",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": {
      "bool": {
        "filter": {
          "bool": {
            "should": [
              {
                "bool": {
                  "must": [
                    {
                      "exists": {
                        "field": "sections",
                      },
                    },
                    {
                      "term": {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              {
                "term": {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "minimum_should_match": 0,
        "must": [],
        "should": {
          "multi_match": {
            "fields": [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.email",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": [
      "_score",
      {
        "class.classId.keyword": {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
]
`;

exports[`searcher generateQuery generates a query without filters 1`] = `
{
  "aggregations": {
    "nupath": {
      "terms": {
        "field": "class.nupath.keyword",
        "size": 1000,
      },
    },
  },
  "from": 0,
  "query": {
    "bool": {
      "filter": {
        "bool": {
          "should": [
            {
              "bool": {
                "must": [
                  {
                    "exists": {
                      "field": "sections",
                    },
                  },
                  {
                    "term": {
                      "class.termId": "202030",
                    },
                  },
                ],
              },
            },
            {
              "term": {
                "type": "employee",
              },
            },
          ],
        },
      },
      "minimum_should_match": 0,
      "must": [],
      "should": {
        "multi_match": {
          "fields": [
            "class.name^2",
            "class.name.autocomplete",
            "class.subject^4",
            "class.classId^3",
            "sections.profs",
            "sections.crn",
            "employee.name^2",
            "employee.email",
            "employee.phone",
          ],
          "query": "fundies",
          "type": "most_fields",
        },
      },
    },
  },
  "size": 10,
  "sort": [
    "_score",
    {
      "class.classId.keyword": {
        "order": "asc",
        "unmapped_type": "keyword",
      },
    },
  ],
}
`;
